File: app-android/app/src/main/java/de/tutao/tutanota/alarms/AlarmNotification.java
package de.tutao.tutanota.alarms;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.room.Embedded;
import androidx.room.Entity;
import androidx.room.TypeConverter;
import androidx.room.TypeConverters;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Collection;
import java.util.Date;
import java.util.Objects;

import de.tutao.tutanota.Crypto;
import de.tutao.tutanota.CryptoError;
import de.tutao.tutanota.EncryptionUtils;
import de.tutao.tutanota.IdTuple;
import de.tutao.tutanota.OperationType;

@Entity(primaryKeys = "identifier")
@TypeConverters(AlarmNotification.OperationTypeConverter.class)
public class AlarmNotification {
	private OperationType operation;
	private String summary;
	private String eventStart;
	private String eventEnd;
	@Embedded
	@NonNull
	private AlarmInfo alarmInfo;
	@Embedded
	private RepeatRule repeatRule;
	@Embedded(prefix = "key")
	private NotificationSessionKey notificationSessionKey;
	private String user;

	public AlarmNotification(OperationType operation, String summary, String eventStart, String eventEnd,
							 AlarmInfo alarmInfo,
							 RepeatRule repeatRule,
							 @Nullable NotificationSessionKey notificationSessionKey, // in case of a delete operation there is no session key
							 String user) {
		this.operation = operation;
		this.summary = summary;
		this.eventStart = eventStart;
		this.eventEnd = eventEnd;
		this.alarmInfo = alarmInfo;
		this.repeatRule = repeatRule;
		this.notificationSessionKey = notificationSessionKey;
		this.user = user;
	}

	public static AlarmNotification fromJson(JSONObject jsonObject, Collection<String> pushIdentifierIds) throws JSONException {
		OperationType operationType = OperationType.values()[jsonObject.getInt("operation")];
		String summaryEnc = jsonObject.getString("summary");
		String eventStartEnc = jsonObject.getString("eventStart");
		String eventEndEnc = jsonObject.getString("eventEnd");
		RepeatRule repeatRule;
		if (jsonObject.isNull("repeatRule")) {
			repeatRule = null;
		} else {
			repeatRule = RepeatRule.fromJson(jsonObject.getJSONObject("repeatRule"));
		}
		AlarmInfo alarmInfo = AlarmInfo.fromJson(jsonObject.getJSONObject("alarmInfo"));
		JSONArray notificationSessionKeysJSON = jsonObject.getJSONArray("notificationSessionKeys");

		NotificationSessionKey notificationSessionKey = null;
		if (notificationSessionKeysJSON.length() == 1) {
			notificationSessionKey = NotificationSessionKey.fromJson(notificationSessionKeysJSON.getJSONObject(0));
		} else {
			for (int i = 0; i < notificationSessionKeysJSON.length(); i++) {
				NotificationSessionKey sessionKey = NotificationSessionKey.fromJson(notificationSessionKeysJSON.getJSONObject(i));
				if (pushIdentifierIds.contains(sessionKey.getPushIdentifier().getElementId())) {
					notificationSessionKey = sessionKey;
					break;
				}
			}
		}

		String user = jsonObject.getString("user");
		return new AlarmNotification(operationType, summaryEnc, eventStartEnc, eventEndEnc, alarmInfo,
				repeatRule, notificationSessionKey, user);
	}

	public OperationType getOperation() {
		return operation;
	}

	public Date getEventStartDec(Crypto crypto, byte[] sessionKey) throws CryptoError {
		return EncryptionUtils.decryptDate(this.eventStart, crypto, sessionKey);

	}

	public Date getEventEndDec(Crypto crypto, byte[] sessionKey) throws CryptoError {
		return EncryptionUtils.decryptDate(this.eventEnd, crypto, sessionKey);

	}

	public String getSummaryDec(Crypto crypto, byte[] sessionKey) throws CryptoError {
		return EncryptionUtils.decryptString(this.summary, crypto, sessionKey);
	}

	public RepeatRule getRepeatRule() {
		return repeatRule;
	}

	public AlarmInfo getAlarmInfo() {
		return alarmInfo;
	}

	public String getSummary() {
		return summary;
	}

	public String getEventStart() {
		return eventStart;
	}

	public String getEventEnd() {
		return eventEnd;
	}

	@Nullable
	public NotificationSessionKey getNotificationSessionKey() {
		return notificationSessionKey;
	}

	public String getUser() {
		return user;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o)
			return true;
		if (o == null || getClass() != o.getClass())
			return false;
		AlarmNotification that = (AlarmNotification) o;
		return Objects.equals(alarmInfo.getIdentifier(), that.alarmInfo.getIdentifier());
	}

	@Override
	public int hashCode() {
		return Objects.hash(alarmInfo.getIdentifier());
	}

	public static class NotificationSessionKey {
		@Embedded
		private final IdTuple pushIdentifier;
		private final String pushIdentifierSessionEncSessionKey;


		public static NotificationSessionKey fromJson(JSONObject jsonObject) throws JSONException {
			JSONArray id = jsonObject.getJSONArray("pushIdentifier");
			return new NotificationSessionKey(
					new IdTuple(id.getString(0), id.getString(1)),
					jsonObject.getString("pushIdentifierSessionEncSessionKey")
			);
		}

		public NotificationSessionKey(IdTuple pushIdentifier,
									  String pushIdentifierSessionEncSessionKey) {
			this.pushIdentifier = pushIdentifier;
			this.pushIdentifierSessionEncSessionKey = pushIdentifierSessionEncSessionKey;
		}

		public IdTuple getPushIdentifier() {
			return pushIdentifier;
		}

		public String getPushIdentifierSessionEncSessionKey() {
			return pushIdentifierSessionEncSessionKey;
		}
	}

	public static class OperationTypeConverter {
		@TypeConverter
		public OperationType fromNumber(int number) {
			return OperationType.values()[number];
		}

		@TypeConverter
		public int numberToOperationType(OperationType operationType) {
			return operationType.ordinal();
		}
	}
}

File: app-android/app/src/main/java/de/tutao/tutanota/alarms/AlarmNotification.kt
package de.tutao.tutanota.alarms

import androidx.room.Embedded
import androidx.room.Entity
import androidx.room.TypeConverter
import androidx.room.TypeConverters
import de.tutao.tutanota.*
import de.tutao.tutanota.alarms.AlarmNotification.OperationTypeConverter
import org.json.JSONException
import org.json.JSONObject
import java.util.*

@Entity(primaryKeys = ["identifier"])
@TypeConverters(OperationTypeConverter::class)
class AlarmNotification(
	val operation: OperationType,
	val summary: String,
	val eventStart: String,
	val eventEnd: String,
	@field:Embedded val alarmInfo: AlarmInfo,
	@field:Embedded val repeatRule: RepeatRule?,
		// in case of a delete operation there is no session key
	@field:Embedded(prefix = "key") val notificationSessionKey: NotificationSessionKey?,
	val user: String,
) {

	@Throws(CryptoError::class)
	fun getEventStartDec(crypto: Crypto, sessionKey: ByteArray): Date {
		return EncryptionUtils.decryptDate(eventStart, crypto, sessionKey)
	}

	@Throws(CryptoError::class)
	fun getEventEndDec(crypto: Crypto, sessionKey: ByteArray): Date {
		return EncryptionUtils.decryptDate(eventEnd, crypto, sessionKey)
	}

	@Throws(CryptoError::class)
	fun getSummaryDec(crypto: Crypto, sessionKey: ByteArray): String {
		return EncryptionUtils.decryptString(summary, crypto, sessionKey)
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other == null || javaClass != other.javaClass) return false
		val that = other as AlarmNotification
		return alarmInfo.identifier == that.alarmInfo.identifier
	}

	override fun hashCode(): Int {
		return Objects.hash(alarmInfo.identifier)
	}

	class NotificationSessionKey(
		@field:Embedded val pushIdentifier: IdTuple,
		val pushIdentifierSessionEncSessionKey: String,
	) {

		companion object {
			@Throws(JSONException::class)
			fun fromJson(jsonObject: JSONObject): NotificationSessionKey {
				val id = jsonObject.getJSONArray("pushIdentifier")
				return NotificationSessionKey(
						IdTuple(id.getString(0), id.getString(1)),
						jsonObject.getString("pushIdentifierSessionEncSessionKey")
				)
			}
		}
	}

	class OperationTypeConverter {
		@TypeConverter
		fun fromNumber(number: Int): OperationType {
			return OperationType.values()[number]
		}

		@TypeConverter
		fun numberToOperationType(operationType: OperationType): Int {
			return operationType.ordinal
		}
	}

	companion object {
		@Throws(JSONException::class)
		fun fromJson(jsonObject: JSONObject, pushIdentifierIds: Collection<String?>): AlarmNotification {
			val operationType = OperationType.values()[jsonObject.getInt("operation")]
			val summaryEnc = jsonObject.getString("summary")
			val eventStartEnc = jsonObject.getString("eventStart")
			val eventEndEnc = jsonObject.getString("eventEnd")
			val repeatRule: RepeatRule? = if (jsonObject.isNull("repeatRule")) {
				null
			} else {
				RepeatRule.fromJson(jsonObject.getJSONObject("repeatRule"))
			}
			val alarmInfo: AlarmInfo = AlarmInfo.fromJson(jsonObject.getJSONObject("alarmInfo"))
			val notificationSessionKeysJSON = jsonObject.getJSONArray("notificationSessionKeys")
			var notificationSessionKey: NotificationSessionKey? = null
			if (notificationSessionKeysJSON.length() == 1) {
				notificationSessionKey = NotificationSessionKey.fromJson(notificationSessionKeysJSON.getJSONObject(0))
			} else {
				for (i in 0 until notificationSessionKeysJSON.length()) {
					val sessionKey = NotificationSessionKey.fromJson(notificationSessionKeysJSON.getJSONObject(i))
					if (pushIdentifierIds.contains(sessionKey.pushIdentifier.elementId)) {
						notificationSessionKey = sessionKey
						break
					}
				}
			}
			val user = jsonObject.getString("user")
			return AlarmNotification(operationType, summaryEnc, eventStartEnc, eventEndEnc, alarmInfo,
					repeatRule, notificationSessionKey, user)
		}
	}
}